packages/flutter/lib/src/painting/inline_span.dart

```diff
@@ -0,0 +1,257 @@
+  ///   }
+  ///
+  ///   @override
+  ///   void dispose() {
+  ///     _longPressRecognizer.dispose();
+  ///     super.dispose();
+  ///   }
+  ///
+  ///   void _handlePress() {
+  ///     HapticFeedback.vibrate();
+  ///   }
+  ///
+  ///   @override
+  ///   Widget build(BuildContext context) {
+  ///     return RichText(
+  ///       text: TextSpan(
+  ///         text: 'Can you ',
+  ///         style: TextStyle(color: Colors.black),
+  ///         children: <InlineSpan>[
+  ///           TextSpan(
+  ///             text: 'find the',
+  ///             style: TextStyle(
+  ///               color: Colors.green,
+  ///               decoration: TextDecoration.underline,
+  ///               decorationStyle: TextDecorationStyle.wavy,
+  ///             ),
+  ///             recognizer: _longPressRecognizer,
+  ///           ),
+  ///           TextSpan(
+  ///             text: ' secret?',
+  ///           ),
+  ///         ],
+  ///       ),
+  ///     );
+  ///   }
+  /// }
+  /// ```
+  /// {@end-tool}
+  final GestureRecognizer recognizer;
+
+  /// An alternative semantics label for this text.
+  ///
+  /// If present, the semantics of this span will contain this value instead
+  /// of the actual text.
+  ///
+  /// This is useful for replacing abbreviations or shorthands with the full
+  /// text value:
+  ///
+  /// ```dart
+  /// TextSpan(text: r'$$', semanticsLabel: 'Double dollars')
+  /// ```
+  final String semanticsLabel;
+
+  /// Utility method to convert InlineSpan into [TextSpan] or [WidgetSpan].
+  ///
+  /// Will return null if span is not of type T or null. Otherwise, return
+  /// the span cast as T. Values of T that are not subclasses of [InlineSpan]
+  /// will result in null.
+  static T asType<T>(InlineSpan span) {
+    if (span is! T)
+      return null;
+    return span as T;
+  }
+
+  /// Apply the properties of this object to the given [ParagraphBuilder], from
+  /// which a [Paragraph] can be obtained. [Paragraph] objects can be drawn on
+  /// [Canvas] objects.
+  void build(ui.ParagraphBuilder builder, { double textScaleFactor = 1.0, List<PlaceholderDimensions> dimensions });
+
+  /// Walks this InlineSpan and any descendants in pre-order and calls [visitor]
+  /// for each span that has content.
+  ///
+  /// When [visitor] returns true, the walk will continue. When [visitor] returns
+  /// false, then the walk will end.
+  ///
+  /// The type of the span may be checked using [asType].
+  bool visitInlineSpan(bool visitor(InlineSpan span));
+
+  /// Returns the text span that contains the given position in the text.
+  InlineSpan getSpanForPosition(TextPosition position);
+
+  /// Flattens the [InlineSpan] tree into a single string.
+  ///
+  /// Styles are not honored in this process.
+  String toPlainText();
+
+  /// Returns the UTF-16 code unit at the given index in the flattened string.
+  ///
+  /// Returns null if the index is out of bounds.
+  int codeUnitAt(int index);
```

Comment:
what is this used for? this seems like a terrible terrible api...

Reply:
This is part of the existing API for TextSpan. I can move this into TextSpan, and do type checking in the places this is called.

Reply:
What's it used for though? "Code units" aren't an API that we should be exposing at all, ideally.

Reply:
https://github.com/flutter/flutter/pull/7432 is the source of the API. Looks like we use it to ensure we are not at a textposition in between a UTF16 codepoint.

Reply:
maybe we should replace this with something like "isInSurrogate" or "isValidLocation"?